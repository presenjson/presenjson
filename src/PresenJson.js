import React, { Component} from 'react';
import PropTypes from 'prop-types';
import R from 'ramda';
import cx from 'classnames';
import Layer from './Layer';

class PresenJson extends Component {
    state = {
        paused: !this.props.autoPlay,
        initial: true
    }

    togglePlayback = () => {
        this.setState({ paused: !this.state.paused, initial: false });
    }

    render() {
        const classNames = cx({
            presenjson: true,
            paused: this.state.paused && !this.state.initial,
            initial: this.state.initial
        });
        const layers = R.flatten(R.of(this.props.children));

        return (<div className={classNames} onClick={this.togglePlayback}>
            <div className='layers'>
                {layers.map((track, i) => <Layer
                    {...track.props}
                    paused={this.state.paused}
                    startAt={this.props.startAt}
                    key={i} />)}
            </div>
            <div className='playback-state' />
        </div>);
    }
}

PresenJson.defaultProps = {
    startAt: 0,
    debug: false,
    autoPlay: false
};

export default PresenJson;
