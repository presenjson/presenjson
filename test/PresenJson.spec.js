import React from 'react';
import { shallow } from 'enzyme';
import PresenJson from '../src/PresenJson';

describe('PresenJson', () => {
    it('renders poster', () => {
        const Poster = () => <div />;
        const presenjson = shallow(<PresenJson poster={Poster} />);
        expect(presenjson).toMatchSnapshot();
    });
    it('renders all layers', () => {
        const Track = () => false;
        const presenjson = shallow(
            <PresenJson startAt={1000} data="data">
                <Track some="prop">
                    <div />
                </Track>
                <Track other="prop">
                    <div />
                </Track>
            </PresenJson>
        );
        expect(presenjson).toMatchSnapshot();
    });
    it('renders only solo layers', () => {
        const Track = () => false;
        const presenjson = shallow(
            <PresenJson startAt={1000} data="data">
                <Track some="prop" solo>
                    <div />
                </Track>
                <Track other="prop">
                    <div />
                </Track>
            </PresenJson>
        );
        expect(presenjson).toMatchSnapshot();
    });
    describe('componentDidMount', () => {
        it('sets initial, paused state', () => {
            const presenJson = new PresenJson({});
            presenJson.setState = jest.fn();

            presenJson.componentDidMount();

            expect(presenJson.setState).toHaveBeenCalledWith({
                paused: true,
                initial: true
            });
        });
        it('sets initial, playing state when autoPlay is set', () => {
            const presenJson = new PresenJson({ autoPlay: true });
            presenJson.setState = jest.fn();

            presenJson.componentDidMount();

            expect(presenJson.setState).toHaveBeenCalledWith({
                paused: false,
                initial: true
            });
        });
        it('sets initial, playing state when play is set', () => {
            const presenJson = new PresenJson({ play: true });
            presenJson.setState = jest.fn();

            presenJson.componentDidMount();

            expect(presenJson.setState).toHaveBeenCalledWith({
                paused: false,
                initial: false
            });
        });
    });
    describe('componentWillReceiveProps', () => {
        it('pauses if props.play is false and changes', () => {
            const presenJson = new PresenJson({ play: false });
            presenJson.setState = jest.fn();

            presenJson.componentWillReceiveProps({ play: true });

            expect(presenJson.setState).toHaveBeenCalledWith({
                paused: false,
                initial: false
            });
        });
    });
    describe('togglePlayback', () => {
        it('toggles playback', () => {
            const presenJson = new PresenJson({ play: true });
            presenJson.setState = jest.fn();

            presenJson.togglePlayback();

            expect(presenJson.setState).toHaveBeenCalledWith({
                paused: true,
                initial: false
            });
        });
    });
    describe('onLoad', () => {
        it('triggers props onLoad with longest length when all rendered layers have called onLoad', () => {
            const onLoad = jest.fn();
            const presenJson = new PresenJson({ onLoad });
            presenJson.layersToRender = [{}, {}, {}];

            presenJson.onLoad(1000);
            expect(onLoad).not.toHaveBeenCalled();

            presenJson.onLoad(10000);
            expect(onLoad).not.toHaveBeenCalled();

            presenJson.onLoad(2000);
            expect(onLoad).toHaveBeenCalledWith(10000);
        });
    });
});
