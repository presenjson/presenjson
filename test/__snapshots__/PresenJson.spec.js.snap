// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PresenJson renders all layers 1`] = `
<div
  className="presenjson initial"
  onClick={[Function]}
>
  <div
    className="layers"
  >
    <Layer
      data="data"
      key="0"
      layer={0}
      onLoad={[Function]}
      paused={true}
      some="prop"
      startAt={1000}
    >
      <div />
    </Layer>
    <Layer
      data="data"
      key="1"
      layer={1}
      onLoad={[Function]}
      other="prop"
      paused={true}
      startAt={1000}
    >
      <div />
    </Layer>
  </div>
  <div
    className="playback-state"
  />
</div>
`;

exports[`PresenJson renders only solo layers 1`] = `
<div
  className="presenjson initial"
  onClick={[Function]}
>
  <div
    className="layers"
  >
    <Layer
      data="data"
      key="0"
      layer={0}
      onLoad={[Function]}
      paused={true}
      solo={true}
      some="prop"
      startAt={1000}
    >
      <div />
    </Layer>
  </div>
  <div
    className="playback-state"
  />
</div>
`;

exports[`PresenJson renders poster 1`] = `
<div
  className="presenjson initial"
  onClick={[Function]}
>
  <div
    className="layers"
  />
  <div
    className="poster"
  >
    <Poster
      autoPlay={false}
      data={Object {}}
      debug={false}
      onLoad={[Function]}
      poster={[Function]}
      startAt={0}
    />
  </div>
  <div
    className="playback-state"
  />
</div>
`;
